File: platform/polyglot/app_yaml_overwrites/mjs/src/logger.ts

/**
 * Unified Logger Interface for App Yaml Overwrites.
 */
export interface ILogger {
    debug(message: string, data?: Record<string, any>): void;
    info(message: string, data?: Record<string, any>): void;
    warn(message: string, data?: Record<string, any>): void;
    error(message: string, data?: Record<string, any>): void;
    trace(message: string, data?: Record<string, any>): void;
}

class ConsoleLogger implements ILogger {
    private prefix: string;
    private currentLevel: number;
    private levels: Record<string, number> = {
        'trace': 0,
        'debug': 1,
        'info': 2,
        'warn': 3,
        'error': 4
    };

    constructor(packageName: string, filename: string) {
        this.prefix = `[${packageName}:${filename}]`;
        const envLevel = (process.env.LOG_LEVEL || 'info').toLowerCase();
        this.currentLevel = this.levels[envLevel] ?? 2;
    }

    private log(level: string, message: string, data?: Record<string, any>) {
        if (this.levels[level] < this.currentLevel) return;

        const timestamp = new Date().toISOString();
        const dataStr = data ? ` ${JSON.stringify(data)}` : '';
        const output = `${this.prefix} ${level.toUpperCase()}: ${message}${dataStr}`;

        switch (level) {
            case 'error':
                console.error(output);
                break;
            case 'warn':
                console.warn(output);
                break;
            default:
                console.log(output);
                break;
        }
    }

    debug(message: string, data?: Record<string, any>): void {
        this.log('debug', message, data);
    }

    info(message: string, data?: Record<string, any>): void {
        this.log('info', message, data);
    }

    warn(message: string, data?: Record<string, any>): void {
        this.log('warn', message, data);
    }

    error(message: string, data?: Record<string, any>): void {
        this.log('error', message, data);
    }

    trace(message: string, data?: Record<string, any>): void {
        this.log('trace', message, data);
    }
}

/**
 * Factory to create a logger instance.
 */
export function create(packageName: string, filename: string): ILogger {
    return new ConsoleLogger(packageName, filename);
}

File: platform/polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/logger.py

from typing import Protocol, Any, runtime_checkable
import sys
import json
from datetime import datetime
import os

@runtime_checkable
class ILogger(Protocol):
    """
    Unified Logger Protocol for App Yaml Overwrites.
    Expected output format: [package:filename] LEVEL: message
    """
    def debug(self, message: str, **kwargs: Any) -> None: ...
    def info(self, message: str, **kwargs: Any) -> None: ...
    def warn(self, message: str, **kwargs: Any) -> None: ...
    def error(self, message: str, **kwargs: Any) -> None: ...
    def trace(self, message: str, **kwargs: Any) -> None: ...

class ConsoleLogger:
    """
    Default implementation of ILogger using standard print (sdtout/stderr).
    Mimics the interface required by the plan.
    """
    def __init__(self, package_name: str, filename: str):
        self.prefix = f"[{package_name}:{filename}]"
        # Simple level mapping
        self.levels = {'trace': 0, 'debug': 1, 'info': 2, 'warn': 3, 'error': 4}
        self.current_level = self.levels.get(os.environ.get('LOG_LEVEL', 'info').lower(), 2)

    def _log(self, level_name: str, message: str, **kwargs: Any):
        if self.levels[level_name] < self.current_level:
            return

        timestamp = datetime.now().isoformat()
        # Format: [package:filename] LEVEL: message {data}
        # If kwargs are present, append them as JSON string
        extra = f" {json.dumps(kwargs)}" if kwargs else ""
        log_line = f"{self.prefix} {level_name.upper()}: {message}{extra}"
        
        # In a real app we might use sys.stderr for error/warn
        print(log_line, file=sys.stdout, flush=True)

    def debug(self, message: str, **kwargs: Any) -> None:
        self._log('debug', message, **kwargs)

    def info(self, message: str, **kwargs: Any) -> None:
        self._log('info', message, **kwargs)

    def warn(self, message: str, **kwargs: Any) -> None:
        self._log('warn', message, **kwargs)

    def error(self, message: str, **kwargs: Any) -> None:
        self._log('error', message, **kwargs)

    def trace(self, message: str, **kwargs: Any) -> None:
        self._log('trace', message, **kwargs)

def create(package_name: str, filename: str) -> ILogger:
    """
    Factory to create a logger instance with standardized context.
    """
    return ConsoleLogger(package_name, filename)
